core/block_svg.js

```diff
@@ -684,6 +684,23 @@ Blockly.BlockSvg.prototype.duplicateAndDragCallback_ = function() {
       // Using domToBlock instead of domToWorkspace means that the new block
       // will be placed at position (0, 0) in main workspace units.
       var newBlock = Blockly.Xml.domToBlock(xml, ws);
+
+      // Scratch-specific: Give shadow dom new IDs to prevent duplicating on paste
```

Comment:
Make this a helper function and call it from here and the copy paste case.  Both because it's good form and because this function is already hard to understand from the coordinate work.